### ISSUE TITLE
How to save checkpoint and resume?

### ISSUE THREAD

**Issue by AykeeSalazar** [ultralytics/yolov5#6891]
So we have to train with 18k images with 100 epoch, it takes 5 hours to complete atleast 10 epoch. When we try to --resume the best.pt for the next one (in another google colab, because of the limit) it says that there is no checkpoint? Can u please help u with this? Thank you

**Comment by Transigent** [comment]
I always specify to save checkpoints on GDrive, so even if training hangs or the session is disconnected, you still have best.pt and last.pt on your GDrive. 

Immediately after first connecting to the Colab session I run this, which connects to my GDrive:
```
from google.colab import drive
drive.mount('/content/gdrive')
```

Heres my training string when training from scratch:

`!python train.py --img 640 --batch 47 --epochs 180 --data '/content/data.yaml' --cfg "/content/yolov5/models/custom_yolov5l.yaml" --weights yolov5l.pt --project "/content/gdrive/MyDrive/Runs"  --name Run_1Mar22 --workers 6   # --cache `

This section
`--project "/content/gdrive/MyDrive/Runs"`
tells it to save checkpoints on the specified folder on GDrive

This bit is the name of a subfolder which stores all the weights and generated files for the training session
`--name Run_1Mar22`

Later, if it is stopped prematurely and I need to resume, I use

```
%%time
%cd /content/yolov5/
!python train.py --resume "/content/gdrive/MyDrive/Runs/Run_1Mar22/weights/last.pt"
```

I specify the location to last.pt in that folder on GDrive. So no matter if you have started another Colab session or not, you can still access the checkpoint location on GDrive, so long as you have connected to GDrive at the start of the new session.

**Comment by Transigent** [comment]
You would also be better off with Colab pro at $10 USD per month or Colab Pro +. You will get better GPUs like P100 or even V100, with P100 my 25k image data set takes about 40mins per epoch. With V100 its 18 minutes. on Colab Pro+ I get A100 GPUs on occasion, for 6 minutes per epoch. You get 40GB on A100 as opposed to 16GB on the others, which means your batch can be much higher.

**Comment by glenn-jocher** [comment]
@AykeeSalazar 👋 Hello! Thanks for asking about **resuming training**. YOLOv5 🚀 Learning Rate (LR) [schedulers](https://github.com/ultralytics/yolov5/blob/9f3a388cea66ca71b9bbe057929eed8fbc920a7e/train.py#L109-L142) follow predefined LR curves for the fixed number of `--epochs` defined at training start (`default=300`), and are designed to fall to a minimum LR on the final epoch for best training results. For this reason you can not modify the number of epochs once training has started.

<img width="850" alt="LR Curves" src="https://user-images.githubusercontent.com/26833433/119332195-8e113a80-bc88-11eb-8eb7-7494ff1d5c01.png">

If your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.run --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck 🍀 and let us know if you have any other questions!

**Comment by AykeeSalazar** [comment]
> I always specify to save checkpoints on GDrive, so even if training hangs or the session is disconnected, you still have best.pt and last.pt on your GDrive.
> 
> Immediately after first connecting to the Colab session I run this, which connects to my GDrive:
> 
> ```
> from google.colab import drive
> drive.mount('/content/gdrive')
> ```
> 
> Heres my training string when training from scratch:
> 
> `!python train.py --img 640 --batch 47 --epochs 180 --data '/content/data.yaml' --cfg "/content/yolov5/models/custom_yolov5l.yaml" --weights yolov5l.pt --project "/content/gdrive/MyDrive/Runs" --name Run_1Mar22 --workers 6 # --cache `
> 
> This section `--project "/content/gdrive/MyDrive/Runs"` tells it to save checkpoints on the specified folder on GDrive
> 
> This bit is the name of a subfolder which stores all the weights and generated files for the training session `--name Run_1Mar22`
> 
> Later, if it is stopped prematurely and I need to resume, I use
> 
> ```
> %%time
> %cd /content/yolov5/
> !python train.py --resume "/content/gdrive/MyDrive/Runs/Run_1Mar22/weights/last.pt"
> ```
> 
> I specify the location to last.pt in that folder on GDrive. So no matter if you have started another Colab session or not, you can still access the checkpoint location on GDrive, so long as you have connected to GDrive at the start of the new session.

Hi yes, that is the first thing I did. But unfortunately, after doing so with 100 epochs with different runtime (because of the limited time Coab gas to offer), the model cannot detect anything. I keep on running the same code but with the latest "last. pt". But I figured it now, I just need to put the runs folder every time into my drive and to use --resume. 

Sadly, Google Colab Pro is not yet available here in the Philippines.

**Comment by nurstd** [comment]
> I always specify to save checkpoints on GDrive, so even if training hangs or the session is disconnected, you still have best.pt and last.pt on your GDrive.
> 
> Immediately after first connecting to the Colab session I run this, which connects to my GDrive:
> 
> ```
> from google.colab import drive
> drive.mount('/content/gdrive')
> ```
> 
> Heres my training string when training from scratch:
> 
> `!python train.py --img 640 --batch 47 --epochs 180 --data '/content/data.yaml' --cfg "/content/yolov5/models/custom_yolov5l.yaml" --weights yolov5l.pt --project "/content/gdrive/MyDrive/Runs" --name Run_1Mar22 --workers 6 # --cache `
> 
> This section `--project "/content/gdrive/MyDrive/Runs"` tells it to save checkpoints on the specified folder on GDrive
> 
> This bit is the name of a subfolder which stores all the weights and generated files for the training session `--name Run_1Mar22`
> 
> Later, if it is stopped prematurely and I need to resume, I use
> 
> ```
> %%time
> %cd /content/yolov5/
> !python train.py --resume "/content/gdrive/MyDrive/Runs/Run_1Mar22/weights/last.pt"
> ```
> 
> I specify the location to last.pt in that folder on GDrive. So no matter if you have started another Colab session or not, you can still access the checkpoint location on GDrive, so long as you have connected to GDrive at the start of the new session.

how can i continue training to 5000 epochs if previously i only set training completed in 2000 epochs

**Comment by glenn-jocher** [comment]
@mbahwahid 👋 Hello! Thanks for asking about **resuming training**. YOLOv5 🚀 Learning Rate (LR) [schedulers](https://github.com/ultralytics/yolov5/blob/aa542ce6a65658ff931fee9bbab77c0145c152f0/train.py#L175-L181) follow predefined LR curves for the fixed number of `--epochs` defined at training start (`default=300`), and are designed to fall to a minimum LR on the final epoch for best training results. For this reason you can not modify the number of epochs once training has started.

<img width="850" alt="Screenshot 2022-04-10 at 11 11 51" src="https://user-images.githubusercontent.com/26833433/162611202-4e0b611f-6cca-4e43-94fc-bbc44e6447f8.png">

If your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.run --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck 🍀 and let us know if you have any other questions!
